File: backend/apps/topics/bank.py
"""The built-in bank: the ten genres, the warm-ups, the styles and the files.

The rows themselves are one JSON file per genre under `data/topics/`, the
speak ones carried over from v0 unchanged. Editing a file and running
`seed_topics` is the whole content workflow.

A file is read by the loader its genre's mode names - `load` for prompts,
`load_passages` for tongue twisters - and the two check different things
because they hold different things. Nothing here reads a file without
knowing which kind it is.
"""

import json
from pathlib import Path

from django.utils.text import slugify

TOPICS_DIR = Path(__file__).resolve().parent.parent.parent / "data" / "topics"

#: (slug, name, icon, blurb), in the order the picker shows them. Ten and no
#: more: a flat scroll list stops working somewhere around fifteen, which is
#: why ten needs no shelves.
GENRES: tuple[tuple[str, str, str, str], ...] = (
    ("general", "General", "dices", "Ordinary things, meals, journeys. A bit of everything - start here."),
    ("relationships", "Relationships", "heart", "Friends, family, love, and the awkward bits."),
    ("career", "Career & work", "briefcase", "Work, ambition, focus, and the questions interviewers ask."),
    ("money-business", "Money & business", "banknote", "Money, building things, and selling them."),
    ("tech-ai", "Tech & AI", "bot", "Machines, models, and what they are doing to us."),
    ("science", "Science & climate", "microscope", "How the world works, and the arguments about the planet."),
    ("health", "Health & mind", "dumbbell", "Bodies, attention, rest, and why people do what they do."),
    (
        "philosophy",
        "Philosophy & ethics",
        "brain",
        "Big questions with no right answer, and hard choices with no clean side.",
    ),
    ("culture", "Culture", "clapperboard", "History, screens, and the hypotheticals you have argued at 2am."),
    (
        "deep-research",
        "Deep research",
        "graduation-cap",
        "Quantum, economics, machine learning. Hard ideas said out loud, and the one genre worth a long prep.",
    ),
)

#: How a genre is practised. `speak` is every genre above: a prompt you are
#: asked to talk about. `read` is a warm-up: text you read aloud off a
#: scroller, which skips prep (there is nothing to think about) and has no
#: clock (the scroll is the timer). One column and one branch, not a second
#: round (docs/DECISIONS.md).
MODE_SPEAK = "speak"
MODE_READ = "read"

#: The warm-ups, kept out of GENRES on purpose: the ten are a flat unheaded
#: list in the picker and the decision behind ten is not reopened by adding
#: to it. These sit in their own labelled group, the way Yours already does.
WARM_UPS: tuple[tuple[str, str, str, str], ...] = (
    (
        "tongue-twisters",
        "Tongue twisters",
        "mic",
        "Long ones, read aloud off a scroller. Warm up your mouth before you talk.",
    ),
)


def all_genres() -> tuple[tuple[str, str, str, str, str], ...]:
    """Every built-in genre with its mode, in picker order: the ten, then
    the warm-ups. The seeder and the API both walk this rather than GENRES,
    so a warm-up cannot be half-declared."""
    return tuple((*genre, MODE_SPEAK) for genre in GENRES) + tuple((*genre, MODE_READ) for genre in WARM_UPS)


#: (key, label, hint): how you are asked to talk about a topic. Surprise me is
#: first because it is the default, and it is not a style: it means no
#: filter, and it is never stored on a topic. The other four are four modes
#: (open, argue, teach, tell), and that is the test a fifth has to pass.
STYLES: tuple[tuple[str, str, str], ...] = (
    ("surprise", "Surprise me", "Any style. The default."),
    ("just-talk", "Just talk", "An open prompt. Go wherever it takes you."),
    ("hot-take", "Hot take", "Pick a side in the first ten seconds."),
    ("explain", "Explain it simply", "Teach it to someone who has never heard of it."),
    ("story", "Tell a story", "One moment, one scene, one point."),
)

SURPRISE = "surprise"

#: The styles a built-in topic may carry.
STYLE_KEYS: frozenset[str] = frozenset(key for key, *_ in STYLES if key != SURPRISE)

MAX_TEXT = 200
MAX_IMAGE = 500

#: A passage is a paragraph, not a prompt: 100 to 120 words, which is 40 to
#: 60 seconds of reading aloud at the speeds the scroller offers. The
#: ceiling is a guard against a runaway file, not a target.
MAX_PASSAGE = 1200
MIN_PASSAGE = 200

#: What a passage carries in `level`, and the whole of what it carries:
#: difficulty is the only axis a passage has, since you do not choose how
#: to talk about words you read verbatim. Its own vocabulary on its own
#: column on its own table, which is the difference between this and the
#: first cut. Easy first, because the page leads with it.
LEVELS: tuple[tuple[str, str], ...] = (("easy", "Easy"), ("hard", "Hard"))
LEVEL_KEYS: frozenset[str] = frozenset(key for key, _ in LEVELS)


def slugify_topic(text: str) -> str:
    return slugify(text)[:220]


def load(slug: str) -> list[dict] | None:
    """The prompts in a speak genre's file, checked, or None when there is
    no file.

    A missing file is skipped rather than fatal, so a genre can be declared
    before its bank is written. A bad line is fatal: a topic that silently
    lands with no style, or none at all, is far harder to notice than an
    import error.
    """
    payload = _read(slug)
    if payload is None:
        return None
    topics = []
    for item in payload["topics"]:
        text = item["text"].strip()
        style = item.get("style", item.get("format", "just-talk"))
        if not 1 <= len(text) <= MAX_TEXT:
            raise ValueError(f"{slug}.json: text must be 1 to {MAX_TEXT} characters, got {len(text)}: {text[:60]!r}")
        if style not in STYLE_KEYS:
            raise ValueError(f"{slug}.json: unknown style {style!r} on {text[:60]!r}")
        # A picture as well as a sentence: `image` on the row is what makes
        # it a picture topic (docs/DECISIONS.md, 2026-09-09). The key is not
        # required, so every file written before this one still loads.
        image = item.get("image", "").strip()
        if len(image) > MAX_IMAGE:
            raise ValueError(f"{slug}.json: overlong image on {text[:60]!r}")
        # A prompt slugifies from its own words and always has. Only a
        # passage names its own, so a speak file naming one is a file that
        # thinks it is the other kind.
        if str(item.get("slug", "")).strip():
            raise ValueError(f"{slug}.json: only a tongue twister names its own slug")
        topics.append({"text": text, "style": style, "image": image, "slug": slugify_topic(text)})
    return topics


def load_passages(slug: str) -> list[dict] | None:
    """The passages in a warm-up's file, checked, or None when there is no
    file.

    A separate loader rather than a mode on `load`, for the reason the
    table is separate: nothing about checking a paragraph is a special case
    of checking a sentence. The bounds are the feature's, the vocabulary is
    `LEVELS`, there are no pictures, and the slug is named in the file
    because slugifying 700 characters gives 220 characters of the first
    sentence, which is no use in a link somebody is meant to paste.
    """
    payload = _read(slug)
    if payload is None:
        return None
    passages = []
    for item in payload["topics"]:
        text = item["text"].strip()
        if not MIN_PASSAGE <= len(text) <= MAX_PASSAGE:
            raise ValueError(
                f"{slug}.json: a passage must be {MIN_PASSAGE} to {MAX_PASSAGE} characters, "
                f"got {len(text)}: {text[:60]!r}"
            )
        level = item.get("level", "")
        if level not in LEVEL_KEYS:
            raise ValueError(f"{slug}.json: unknown level {level!r} on {text[:60]!r}")
        named = str(item.get("slug", "")).strip()
        if not named:
            raise ValueError(f"{slug}.json: a passage names its own slug, and {text[:60]!r} has none")
        passages.append({"text": text, "level": level, "slug": named})
    return passages


def _read(slug: str) -> dict | None:
    path = TOPICS_DIR / f"{slug}.json"
    if not path.exists():
        return None
    return json.loads(path.read_text(encoding="utf-8"))

File: frontend/lib/bank.ts
/* The bank as the page receives it, and nothing that fetches it.

   This module is imported for its *values* by client components (the
   picker asks whether a genre is a warm-up), so it must not reach
   `lib/api.ts`: that imports `next/headers`, and a client component
   pulling it in stops the build outright with "next/headers ... in the
   Pages Router". The fetchers live in `lib/api.ts` with every other
   server-only call, which is the split `lib/owned.ts` already made
   (docs/DECISIONS.md, 2026-09-09).

   The bank as the page receives it: the built-in genres in picker order,
   every active topic, and the styles for the select. One shape for the
   picker, the reel and the style select. */
/* `mode` is absent on the ten and on anything somebody owns, and "read" on
   a warm-up: a genre whose topics are passages read aloud off a scroller
   rather than prompts to talk about. Absent rather than "speak" on every
   row, for the reason `Topic.image` gives below. */
export type Genre = { slug: string; name: string; icon: string; blurb: string; own?: boolean; mode?: "read" };

export const READ = "read";

/** Whether this genre is a warm-up: read aloud, no prep, no clock. */
export function isRead(genre: Genre | undefined): boolean {
  return genre?.mode === READ;
}

/** Where a feature lives. One URL per feature, named by its slug: home is
    the tool and keeps the genre chip, the question and one button, and
    anything that changes what the round *is* takes a page of its own
    rather than a mode on that screen (docs/DECISIONS.md). */
export function featurePath(slug: string): string {
  return `/${slug}`;
}

/* `image` is what makes a topic a picture topic (docs/DECISIONS.md,
   2026-09-09); `text` stays the prompt in both kinds, so a picture round
   is the same round with a picture over the same sentence. Optional, and
   absent rather than empty on the wire: home ships the whole bank inline
   and an empty key on all thousand rows is 34KB the browser parses for
   the sake of the twenty that use it.

   One shape for what the round draws, because a round is a round: the
   engine, the reel and the pool do not care which table a row came from.
   What they must not do is read one key for two meanings, so a prompt
   carries `style` and a passage carries `level`, each absent on the
   other - the same split the tables took (docs/DECISIONS.md). */
export type Topic = { text: string; genre: string; style?: string; slug: string; image?: string; level?: string };
export type Style = { key: string; label: string; hint: string };
export type Bank = { genres: Genre[]; topics: Topic[]; styles: Style[] };

export const EMPTY_BANK: Bank = { genres: [], topics: [], styles: [] };

/* One warm-up genre and its passages. Fetched when the genre is picked
   rather than shipped with the bank: a passage is 100 to 120 words, and
   home ships the whole bank inline so a respin costs no round trip. That
   promise is about the reel; somebody who has just chosen to read a
   passage aloud for a minute will not notice one request. */
export type Passage = { text: string; slug: string; level: string; words: number };
export type ReadGenre = { slug: string; name: string; icon: string; blurb: string; passages: Passage[] };

/* Grouped as the page shows them: easy first, because somebody who has
   never done this needs an obvious way in. A difficulty the bank does not
   hold is simply absent rather than an empty heading. */
export const LEVELS: { key: string; label: string }[] = [
  { key: "easy", label: "Easy" },
  { key: "hard", label: "Hard" },
];

export function passagesByDifficulty(genre: ReadGenre): { label: string; passages: Passage[] }[] {
  return LEVELS.map(({ key, label }) => ({
    label,
    passages: genre.passages.filter((passage) => passage.level === key),
  })).filter((group) => group.passages.length > 0);
}

export function genreBySlug(bank: Bank, slug: string): Genre | undefined {
  return bank.genres.find((genre) => genre.slug === slug);
}

export function topicsOf(bank: Bank, slug: string): Topic[] {
  return bank.topics.filter((topic) => topic.genre === slug);
}

export function topicCounts(bank: Bank): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const topic of bank.topics) counts[topic.genre] = (counts[topic.genre] ?? 0) + 1;
  return counts;
}

/* A genre's topics grouped under the styles the bank offers, in the bank's
   order, skipping Surprise me (it is a filter, not a style) and any style
   the genre happens not to hold. */
export function topicsByStyle(bank: Bank, slug: string): { style: Style; topics: Topic[] }[] {
  const topics = topicsOf(bank, slug);
  return bank.styles
    .filter((style) => style.key !== "surprise")
    .map((style) => ({ style, topics: topics.filter((topic) => topic.style === style.key) }))
    .filter((group) => group.topics.length > 0);
}
